codegen/src/main/twirl/templates/PlayScalaServer/Router.scala.txt

```diff
@@ -0,0 +1,37 @@
+@*
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ *@
+
+@(service: akka.grpc.gen.scaladsl.Service)
+
+package @service.packageName
+
+import javax.inject._
+
+import scala.concurrent.Future
+
+import akka.http.scaladsl.model.{HttpRequest, HttpResponse}
+import akka.stream.Materializer
+
+import play.api.mvc._
+import play.api.mvc.akkahttp.AkkaHttpHandler
+import play.api.routing.Router
+import play.api.routing.Router.Routes
+
+@@Singleton
+class @{service.name}Router @@Inject()(cc: ControllerComponents, impl: controllers.@{service.name}Impl)(implicit mat: Materializer)
+  extends Router {
+
+  val handler = new AkkaHttpHandler {
+    val h = @{service.name}Handler(impl)
+    override def apply(request: HttpRequest): Future[HttpResponse] = h(request)
+  }
+
+  // could look at @{service.name}.name, but isn't this already covered in the route file?
+  override def routes: Routes = { case _ ⇒ handler }
```

Comment:
I have been thinking about this when integrating it in Lagom. 

In Lagom we don't have a route file, but we need to merge the `routes` defined by the `ServiceDescriptor` with this `routes`.
In that case, the gRPC routes will become a `catch-all` matching. 

On the other hand, I don't think we should use `@{service.name}.name` for that. The user should be able to choose on which path the gRPC will be exposed, much like they do when defining the path in a Play routes file.

Reply:
The [spec](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#requests) mentions: "Some gRPC implementations may allow the Path format shown above to be overridden, but this functionality is strongly discouraged. gRPC does not go out of its way to break users that are using this kind of override, but we do not actively support it, and some functionality (e.g., service config support) will not work when the path is not of the form shown above"

So it might be OK to give the user some control over it, but preferably not too obviously. In fact I think it's a bit of a shame we have to mention it in the `routes`, but there I included it to make it more consistent with the other lines.

Reply:
Interesting. In that case, we could see it the other way around. The `Router` we build for a gRPC handler should bind it to the path as prescribed by the spec (ie: ":path" "/" Service-Name "/" {method name}).  

When plugin it in Play, we should instead compose the routes in the app loader and not add it to the `routes` file. The drawback is that this will be fragile as one could always try to override the path by adding something different to the `routes` file.

Reply:
I'm on the fence about whether to add the gRPC routes to the app loader 'automatically' or via the `routes`: doing it via the `routes` might make things 'less magic' and perhaps eventually allow participating in action composition, which could be nice in the light of 'platform consolidation'. On the other hand somehow adding them automatically (not sure how hard that would be?) may be less 'boilerplatey'. I'd love to hear e.g. @richdougherty's thoughts on this.

Reply:
For the MVP I'm OK if we generate one router class for each gRPC service then let the user put each route into the routes file manually. Maybe go as far as printing out a message to help them.

For v2 we could patch the main routes file and/or generate a `grpc.routes` file. If we generate a `grpc.routes` file then the user could map it as `-> / grpc.Routes` or `-> /my/grpc/path grpc.Routes` from the main `routes` file. (Credit to @marcospereira for this idea.)

One thing that we need to be careful of is that if we have a route like `-> / my.custom.Routes` in the `routes` file then I believe it will catch all paths prefixed with `/` (i.e. every request), so we need to make sure it's the last entry in the `routes` file. If we use a route with a prefix, e.g. `->  /helloworld.GreeterService controllers.GreeterServiceRouter` then it will only match requests prefixed with `/helloworld.GreeterService` so it can probably go anywhere in the `routes` file without causing problems.